controllers/selfnoderemediation_controller.go

```diff
@@ -311,6 +321,116 @@ func (r *SelfNodeRemediationReconciler) remediateWithResourceDeletion(snr *v1alp
 	return ctrl.Result{}, nil
 }
 
+func (r *SelfNodeRemediationReconciler) remediateOutOfServiceTaint(snr *v1alpha1.SelfNodeRemediation) (ctrl.Result, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.NodeOutOfServiceVolumeDetach) {
+		err := errors.New("this remediation is disabled")
+		r.logger.Error(err, "Set `NodeOutOfServiceVolumeDetach` feature gate")
+		return ctrl.Result{}, err
+	}
+
+	node, err := r.getNodeFromSnr(snr)
+	if err != nil {
+		r.logger.Error(err, "failed to get node", "node name", snr.Name)
+		return ctrl.Result{}, err
+	}
+
+	if r.isFencingCompleted(snr) {
+		r.logger.Info("fencing completed, cleaning up")
+		if node.Spec.Unschedulable {
+			node.Spec.Unschedulable = false
+			if err := r.Client.Update(context.Background(), node); err != nil {
+				if apiErrors.IsConflict(err) {
+					return ctrl.Result{RequeueAfter: time.Second}, nil
+				}
+				r.logger.Error(err, "failed to mark node as schedulable")
+				return ctrl.Result{}, err
+			}
+		}
+
+		// wait until NoSchedulable taint was removed
+		if utils.TaintExists(node.Spec.Taints, NodeUnschedulableTaint) {
+			return ctrl.Result{RequeueAfter: time.Second}, nil
+		}
+
+		if !r.isResourceDeletionCompleted(node) {
+			isExpired, timeLeft := r.isResourceDeletionExpired(snr)
+			if !isExpired {
+				return ctrl.Result{RequeueAfter: timeLeft}, nil
+			}
+			return ctrl.Result{}, errors.New("Not ready to delete out-of-service taint")
+		}
+
+		if err := r.removeOutOfServiceTaint(node); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if err := r.removeNoExecuteTaint(node); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if controllerutil.ContainsFinalizer(snr, SNRFinalizer) {
+			if err := r.removeFinalizer(snr); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+
+		return ctrl.Result{}, nil
+	}
```

Comment:
I think this code is 95% duplicated, I suggest refactor it into a method
Something like:
```
if r.isFencingCompleted(snr) {
  return postFencingLogic(node, r.isResourceDeletionCompleted)
}
  ```

Reply:
I did. I completely agree to refactor it to commonize.
And it's moving the verification of the resource deletion before setting the fencing complete and then I refactor it.